pyvista/plotting/plotting.py

```diff
@@ -3867,16 +3891,36 @@ def add_volume(
                 actors.append(a)
             return actors
 
-        if not isinstance(volume, (pyvista.UniformGrid, pyvista.RectilinearGrid)):
+        # Make sure structured grids are not less than 3D
+        # ImageData and RectilinearGrid should be olay as <3D
+        if isinstance(volume, pyvista.StructuredGrid):
+            if any(d < 2 for d in volume.dimensions):
+                raise ValueError('StructuredGrids must be 3D dimensional.')
+
+        if isinstance(volume, pyvista.PolyData):
             raise TypeError(
-                f'Type {type(volume)} not supported for volume rendering at this time. Use `pyvista.UniformGrid` or `pyvista.RectilinearGrid`.'
+                f'Type {type(volume)} not supported for volume rendering as it is not 3D.'
             )
+        elif not isinstance(
+            volume, (pyvista.UniformGrid, pyvista.RectilinearGrid, pyvista.UnstructuredGrid)
+        ):
+            volume = volume.cast_to_unstructured_grid()
+
+        # Override mapper choice for UnstructuredGrid
+        if isinstance(volume, pyvista.UnstructuredGrid):
+            volume = volume.triangulate()  # forcibly triangulate, should be inexpensive
```

Comment:
I'm pretty sure VTK and ParaView use this as it's fairly performant.

The tessellate() filter is expensive. Users can opt-in to that on their own

Reply:
Faster to check and then `triangulate`:

```py
import pyvista as pv


ugrid = pv.UniformGrid(dimensions=(50, 50, 50)).cast_to_unstructured_grid()
ugrid_tet = ugrid.triangulate()


def always_triangulate(ugrid):
    return ugrid.triangulate()


def check_then_triangulate(ugrid):
    if not (ugrid.celltypes == pv.celltype.CellType.TETRA).all():
        return ugrid.triangulate()
    return ugrid

"""
>>> timeit check_then_triangulate(ugrid_tet)
63.7 µs ± 74.9 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

timeit always_triangulate(ugrid_tet)
485 µs ± 4.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> timeit check_then_triangulate(ugrid)
2.21 s ± 34.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

timeit always_triangulate(ugrid)
2.23 s ± 7.96 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

"""
```

Reply:
Good by me, can you add that right quick?

Reply:
```suggestion
            # Unstructured grid must be all tetrahedrals
            if not (volume.celltypes == pv.celltype.CellType.TETRA).all():
                return volume.triangulate()
```